text/0039-service-overhaul.md

```diff
@@ -0,0 +1,462 @@
+***This RFC is dependent on [RFC 45](https://github.com/tytan652/rfcs/blob/protocol_api/text/0045-protocols-in-outputs-api.md).***
+
+# Summary
+
+- Make OBS able to accept third-party service plugins
+- Register services with a unique id rather than a common one
+- A service can be provided with multiple protocols
+- Separate Twitch, Restream and YouTube integration and make them plugins
+
+# Motivation
+
+Actually even if OBS has the service API, developer can't create third-party service plugin because there is no mechanism to list those and show their own property view.
+
+Before in OBS, in the Stream settings page, there was a combobox with two choice:
+- Streaming Services
+- Custom Streaming Server
+
+Each one had his own property view because those two are the two actual registered official services.
+This behavior could enable third-party service plugin to exist.
+
+Nowadays in OBS, this page show the list of services with many new elements showed of hidden depending of the selection (like recommended settings). With also Twitch and Restream OAuth integrations. And no use of the property views provided by `rtmp-services`.
+
+We need to restore the usage of property views and the possibility to create and use service plugin.
+
+It will also provide the ability for some stream services to be able to made their own plugin.
+
+*Since using the service API was not possible with the actual OBS, any breaking change in it (if there are) will not impact any plugin for OBS.*
+
+# Design
+
+## UI
+The Stream settings page and the auto-wizard will only contain a combo box listing registered services with also the "Show All" option. And the property view of the service will be shown under it.
+
+A "common" service is a service that is shown in the default service list. To keep this attributes with the UI, a flag will be added to Service API to allow services to not be in this list, so by default third-party plugins will be shown without the need to click on the "Show all" option.
```

Comment:
The idea here is precisely what it sounds like, that a service is "common" - that is, if a user is looking to stream using OBS, a "common" service is the most likely option they will choose.

This does however bring up a concern - we currently only differentiate between common / not common, curated by us. Outside of that, services are provided in a _very_ lengthy alphabetical list. Introducing the ability for users to install any variety of service plugins could grow this list significantly.

Especially with a plugin manager where a user could discover / install any number of service plugins, being able to easily sift through that list would be crucial. I'm unsure how this could be done though - categories? Groupings based on country? There are a lot of services out there (we provide 89 as it is). I wouldn't be against listing newly-installed services from plugins near the top of the list though.

Reply:
I just made this based on [DDRBoxman comments](https://github.com/obsproject/rfcs/pull/39#discussion_r690007308).

Reply:
Note: every service that are not common will have the flag added.

Reply:
The "common" flag has been a subject of considerable debate. Personally once there's a plugin manager, I would be much happier to distribute some default service plugins and then offer some sort of search for new services to "install" into OBS so that each user essentially curates their own "common" list. Personally I'd like to see the [monolithic services.json be split into individual components](https://github.com/obsproject/obs-studio/discussions/3858) which makes it extremely easy to distribute new service plugins that don't require custom oauth features.